M2/Macaulay2/packages/Msolve.m2

```diff
@@ -0,0 +1,527 @@
+
+newPackage(
+	"Msolve",
+	Version => "1.24.05", 
+    	Date => "July 2024",
+    	Authors => {{Name => "Martin Helmer", 
+		  Email => "mhelmer@ncsu.edu", 
+		  HomePage => "http://martin-helmer.com/"}, {Name => "Mike Stillman", 
+		  Email => "mike@math.cornell.edu", 
+		  HomePage => "https://math.cornell.edu/michael-e-stillman"},{Name => "Anton Leykin", 
+		  Email => "leykin@math.gatech.edu", 
+		  HomePage => "https://antonleykin.math.gatech.edu/"}},
+    	Headline => "An interface to the msolve package (https://msolve.lip6.fr/) which computes Groebner Basis and does real root isolation",
+    	AuxiliaryFiles => true,
+	DebuggingMode => false,       
+	OptionalComponentsPresent => msolvePresent
+	);
+msolveEXEwithOpts:=(opt)->"msolve -v "|opt#"level of verbosity"|" -t "|toString(opt#"number of threads");
+export{
+    "msolveGB",
+    "msolveSaturate",
+    "msolveEliminate",
+    "msolveRUR",
+    "msolveLeadMonomials",
+    "msolveRealSolutions"
+    }
+importFrom_Core { "raw", "rawMatrixReadMsolveFile" }
+
+use'readMsolveOutputFile := true;
+readMsolveOutputFile = method()
+readMsolveOutputFile(Ring,String) := Matrix => (R,mOut) -> if use'readMsolveOutputFile 
+    then map(R, rawMatrixReadMsolveFile(raw R, mOut)) else (
+	msolveOut := get mOut;
+    	moutStrL:=separate(",",first separate("[]]",last separate("[[]",msolveOut)));
+    	--the line below should be replaced by a call to the C-function to parse the string
+    	M2Out:=for s in moutStrL list value(s);
+    	matrix {M2Out};
+    	)
+
+inputOkay=method(TypicalValue=>Boolean);
+inputOkay(Ideal):=I->(
+    R:=ring I;
+    if not instance(R,PolynomialRing) then (print "input must be in a polynomial ring over a field";return false;);
+    kk:=coefficientRing(R);
+    if instance(kk,InexactFieldFamily) then(
+        print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	return false;
+	);
+    if char(kk)>0 then (
+	if (char(kk)<2^16) or (char(kk)>2^32) then(
+	    print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	    return false;
+	    );
+	);
+    return true;
+    );
+msolveGB=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveGB(Ideal):=opt->I->(
+    if not inputOkay(I) then return 0;
+    mIn:=temporaryFileName()|".ms";
+    R:=ring I;
+    kk:=coefficientRing(R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    << "msolve input file is called: " << mIn << endl;
+    << "msolve output file is called: " << mOut << endl;
```

Comment:
These (and ideally also the `callStr` should be printed only if e.g. `debugLevel > 0` and using `printerr`.